### ISSUE TITLE
[Feature] Plan to add `model_register`

### ISSUE THREAD

**Issue by XinDongol** [pytorch/torchtitan#282]
Plan to add `model_register` so users can add their own models

**Comment by wanchaol** [comment]
@XinDongol I think that's a good idea, the way this could work is that `register_model` can insert new entries into various module's module level dict (i.e. [this](https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/__init__.py#L9)), there are couple of things required from user:
1. user registered model must be a meta init ready model (with `init_weights`, `from_model_args`, etc.)
2. user need to provide at least the parallelization method simiar to [parallelize_llama](https://github.com/pytorch/torchtitan/blob/main/torchtitan/parallelisms/parallelize_llama.py#L132) so that torchtitan knows how to shard the model.

In this way any plugged model should be able to run with torchtitan. We don't have time at the moment to implement such a feature, but feel free to submit a PR and we can help review it :)

**Comment by msaroufim** [comment]
@wanchaol @tianyu-l something I was curious to ask about is what do you feel is more inline with torchtitan philosophy
1. Clone the repo, install from source, make local changes and run
2. Install from release, register changes and run

I had this exact same debate thinking about dataset registration https://github.com/pytorch/torchtitan/pull/715

Also curious to hear @carmocca's take since he has tons of experience with this kind of choices

**Comment by tianyu-l** [comment]
I'd say it's closer to 1.
The concern is: The space of pre-training (or even fine-tuning) is a bit large. torchtitan is (so far) thought of as a reference repo, which can't possibly cover all aspects, e.g. in addition to model and dataset, there are optimizer, lr scheduler, etc. Also people may not need all the features, sometimes one can just delete the things they don't want and keep a minimal version of torchtitan, e.g. https://github.com/facebookresearch/lingua. But I guess we don't have to keep the boundary strict?

Another thing is that torchtitan is still constantly evolving, sometimes depending on the latest pytorch nightly. Making a stable release may force use to align with pytorch stable release, or also start nightly release, which adds to the operational costs.

I'm also curious what others think.

**Comment by carmocca** [comment]
Fully agree with @tianyu-l.

To me, the strength of this repo is being correct, fast, and backed by the PyTorch team.

So far from my experience, my changes have been separate and easy to fit into the released code.
The main places that required adaptations were logging (fixed by #699), the logic around the tokenizer/dataloader, and the logic around the optimizer.
From the perspective of registering an "arbitrary" model, the biggest lift for most users will be their parallelize function. Right now the logic is not extensible without forking. It also seems to be regularly changing anyways, which might break people's code that is blind to the insides.

I also thought that lingua was meant to address the usability gap. I have no idea how much torchtitan's team overlap with theirs but it's good to go in two different directions.

Keep doing things the pytorch(tm) way by keeping things generally functional and reusable, then people will be happy.
